### ISSUE TITLE
Migrate to webapp

### ISSUE THREAD

**Issue by sunglim** [beagleterm/beagle-term#60]
http://blog.chromium.org/2016/08/from-chrome-apps-to-web.html

**Comment by jumanjiman** [comment]
fwiw several people at `$dayjob` use beagleterm as chrome app on our chromebooks to connect to network devices. this makes the chromebook a nice alternative to a traditional crash cart.

**Comment by RandomlyKnighted** [comment]
@sunglim what will it take for us to do this migration? I'm not near a computer atm (currently responding on my phone) so I can't view what APIs are missing.

**Comment by sunglim** [comment]
In my personal view, There's no future for Chrome app.

Since 2013 I 've been contributing to Chromium especially extensions and apps. (Internally chrome app and extension they both share same code)
From 2015, some of Chrome extension team guys (more than half) Google. and there's no new feature they implement.  
**I think** after [Chrome Dev Editor](https://www.youtube.com/watch?v=NNLnTz6yIc4), they decided not to focus on Chrome apps. 

It is always better to learn new thing, and get better support. So we need to move forward to webapp.

**Comment by RandomlyKnighted** [comment]
So what would converting this project mean for the project? I know it means continued support by Chrome, but how would this affect using the app within Chrome? Right now I can go to `chrome://apps/` and launch Beagle Term from there. Would this change? Are web apps launched differently? Would our users have to re-learn how to use Beagle Term?

**Comment by sunglim** [comment]
> Would our users have to re-learn how to use Beagle Term?

No. We should provide same UI. there're already many chrome app that just has link to webapp. for example [Google docs](https://chrome.google.com/webstore/detail/google-docs/aohghmighlieiainnegkcijnfilokake?utm_source=chrome-ntp-icon)

I think using serial API from webapp is impossible now, I will take a look at Chromium side. I think some new proposal something like [webUSB ](https://wicg.github.io/webusb/) is necessary.

**Comment by jumanjiman** [comment]
> It is always better to learn new thing rather than holding old technology, and get better support. So we need to move forward to webapp.

understood.

[This post from Google](http://blog.chromium.org/2016/08/from-chrome-apps-to-web.html) seems to imply that apps are still needed on chrome os, but it does not describe the extent to which web apps can work offline or connect to hardware **today**. 

Do you think it will be possible to migrate to web app and continue to work completely offline?

**Comment by RandomlyKnighted** [comment]
I think migrating and being able to work offline might be possible, IF we can get them to add support for the serial API (and possibly the socket API if we implement #70). @jumanjiman In the link to the blog post you provided, Google mentions that we can request prioritizing specific APIs so that our apps will continue to work. [Here is the page](https://docs.google.com/forms/d/e/1FAIpQLSenHdpA8_eqKiVOrWDjWe_KTfJmoSBEqFIh6SMwQ-NRDJnx1Q/viewform?c=0&w=1) where we make that request. I think we should all put in a request for it. I'm going to go ahead and request both APIs be added.

**Comment by jumanjiman** [comment]
@RandomlyKnighted Thanks for the link!

I have submitted a request at https://docs.google.com/forms/d/e/1FAIpQLSenHdpA8_eqKiVOrWDjWe_KTfJmoSBEqFIh6SMwQ-NRDJnx1Q/viewform?c=0&w=1 to add support for the serial API.

**Comment by RandomlyKnighted** [comment]
I asked the Chrome Developers Twitter account if either of these API's were going to be added. In their reply they included the names of the 2 people who decide what API's to support in the new web app. They also asked for a use case so they know why these API's are needed. I provided the use case they wanted. I'll continue talking to them and see if what I can do to get the support added that we need.

**Comment by sunglim** [comment]
I'm asking to Chromium guys directly.


Sungguk Lim <limasdf@gmail.com>
Sep 20 (1 day ago)

to rockot 
Hi Ken!

This is Sungguk. I hope you remember.

Can I ask a question? Recently I saw an article, http://blog.chromium.org/2016/08/from-chrome-apps-to-web.html

So I'm going to migrate my chrome app. but my app uses Chrome.serial api. 

Is it possible to use serial in webapp? do we need to write new proposal something like webUSB? and then implement Chromium side first?

Could you kindly share your thought?

Thanks in advance.

Ken Rockot
4:13 AM (5 hours ago)

to Reilly, me 
I don't think it's possible to migrate chrome.serial usage to a web app at this time. There have been discussions of a Web Serial API in the past but they didn't gain much traction. Might be worth revisiting at some point, but I think a better course of action would be to find a way to use Web USB if possible.

Do you have a sense of how your app is used primarily, i.e. with what kinds of devices? Is it generally for hacking on custom hardware, Arduino, etc? +Reilly, is there a straightforward way to talk to RS232 USB interfaces without the OS interfering?


Reilly Grant <reillyg@chromium.org>
8:44 AM (1 hour ago)

to Ken, me 
The fact that the OS gets in the way is the primary issue blocking use of WebUSB with serial devices. As the Arduino demo shows however it is possible to build a dual-mode device if your microcontroller supports adding additional endpoints.

**Comment by RandomlyKnighted** [comment]
@sunglim Thanks for sharing. So what about devices other non microcontroller applications like interacting with a router or switch? How would you know if the devices supports adding additional endpoints?

**Comment by mdv55** [comment]
Mike here from V&V Systems. We help schools with equipment for their Knowledge Bowl competition. We and the coaches use Beagleterm for communicating with the  Knowledge Bowl timer (which  uses the FTDI FT230X USB UART) for control and display during their practice and competition. Chromebooks are rapidly displacing PCs and Macs in schools here in Washington and surrounding states and Beagleterm is our "go to" solution replacing Windows TeraTerm and native consoles in Mac and Linux -- so we really need a Chrome serial terminal, like Beagleterm, so the coaches statewide can use their Chromebooks for Knowledge Bowl. This is really important to keep working -- let us know if we can help!

**Comment by sunglim** [comment]
@RandomlyKnighted 
Ultimately BeagleTerm should support every RS232 devices(not only devices that support additional endpoint).
IMO, using WebUSB is not the solution. I asked Ken how to start writing new proposal. It must be hard and could be impossible.  let's just see. Who knows? We could be author of newly released proposal such as WebSerial.

**Comment by sunglim** [comment]
@mdv55  Thanks to many contributors(@RandomlyKnighted @jumanjiman and other guys), BeagleTerm can move forward.

**Comment by sunglim** [comment]
news from google
--

Dear Developer,

We’re reaching out to you because you’re registered as a Chrome App developer in the Chrome Web Store. As you might have seen recently, we announced that we will be increasing our focus on the Web and open standards by removing support for Chrome packaged apps and hosted apps on Windows, Mac, and Linux over the next two years. Chrome apps will remain supported and maintained on Chrome OS for the foreseeable future. Chrome extensions will also continue to be supported on all platforms.

Starting in late 2016, newly published Chrome apps will only be available to users on Chrome OS. Existing Chrome apps will remain accessible on all platforms, and you can continue to update them.

In the second half of 2017, the Chrome Web Store will no longer show Chrome apps on Windows, Mac, and Linux, but will continue to surface extensions and themes. In early 2018, users on these platforms will no longer be able to load Chrome apps.

We encourage you to migrate your Chrome app to the open web, and we’ve written a transition guide to help you with this process. If the web platform can’t provide the functionality you need, please let us know why. You can also consider using a Chrome extension or an alternate app platform such as Electron or NW.js.


Best, The Chrome Web Store and Apps Team, 
Google Inc. 1600 Amphitheatre Parkway, Mountain View, CA 94043
MSA Consumer: You've received this mandatory service announcement email to update you on important changes related to the Chrome Web Store and Apps Platform.
